pywbem_mock/_baserepository.py

```diff
@@ -0,0 +1,166 @@
+# version 2.1 of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+#
+# Author: Karl  Schopmeyer <inovadevelopment.com>
+#
+
+
+"""
+Mock support for the WBEMConnection class to allow pywbem users to test the
+pywbem client without requiring a running WBEM server.
+
+For documentation, see mocksupport.rst.
+"""
+from abc import ABCMeta, abstractmethod
+
+import six
+
+from pywbem import CIMClass, CIMQualifierDeclaration, CIMInstance
+
+
+@six.add_metaclass(ABCMeta)
+class BaseObjStore(object):
+
+    def __init__(self, case_insensitive_names, cim_object_type):
+        """
+        """
+        # TODO namespace must point to dynamic object that changes when
+        # namespaces added.
+        self._case_insensitive_names = case_insensitive_names
+        self._cim_object_type = cim_object_type
+
+    @abstractmethod
+    def exists(self, name):
+        """
+        Test if the object defined by name and namespace exists in the object
+        store
+        """
+        pass
+
+    @abstractmethod
+    def get(self, name, copy=True):
+        """
+        Return the object defined by name and namespace if it exists.
+
+        Returns:
+
+        Raises:
+            TODO: What do we return for namespace error.
+            KeyError: Object not in store
+        """
+        pass
+
+    @abstractmethod
+    def create(self, name, object):
+        """
+        """
+        pass
+
+    @abstractmethod
+    def update(self, name, object):
+        """
+        """
+        pass
+
+    @abstractmethod
+    def delete(self, name):
+        """
+        """
+        pass
+
+    def names(self):
+        """
+        Return all of the objects defined in this data store
+
+        Returns:
+            List of objects
+
+        Raises:
+
+        """
+        pass
+
+    @abstractmethod
+    def iter(self):
+        """
+        Return iterator to allow iteration through all the objects of a
+        singe namespace
+
+        Returns:
+            Iterator for a namespace
+
+        Raises:
+
+        """
+        pass
+
+
+@six.add_metaclass(ABCMeta)
+class BaseRepository(object):
+    """
+    An abstract base class for implementing CIM data store for a CIM repository.
+    This class defines the interface that is used by the
+    """
+
+    def __init__(self, repository, obj_store):
+        """
+        Construct the repository by defining the namespace with the supplied
+        default_namespace and setting the objects for the classes, instances,
+        and qualifiers
```

Comment:
* Add a Parameters section.
* The description mentions default_namespace, but it is not clear to me what that is. Also, I think we discussed that at the repository level, there is no concept of a default namespace.
* The repository init argument is a bit confusing. After all, this is the init of a repository.

Reply:
I think you caught me there.  I need to think this one out more.

Reply:
1. Added parameters section to all the API definitons.  Note that We reference the abstract methods from the inmemory repository

2.  TODO: In retrospect the default namespace is an error in that it causes problems if we were to do a repo that loaded from persistent store so that the namespaces already exist. I should remove

3. Remove __init__ from abstract class

4. Namespaces are the top level of the dictionary in the repository rather than a separate list, etc.